sdk/pinocchio/src/cpi.rs

```diff
@@ -13,148 +13,253 @@ use crate::{
+/// number of instruction accounts is not known at compile time. The final number of
+/// accounts passed to the cross-program invocation will be the number of accounts
+/// required by the `instruction`, even if `MAX_ACCOUNTS` is greater than that.
+///
+/// Note that this function is inlined to avoid the overhead of a function call,
+/// but uses stack memory allocation. When a large number of accounts is needed,
+/// it is recommended to use the [`slice_invoke_signed`] function instead to reduce
+/// stack memory utilization.
+///
 /// # Important
 ///
 /// The accounts on the `account_infos` slice must be in the same order as the
-/// `accounts` field of the `instruction`.
-#[inline]
-pub fn slice_invoke_signed(
+/// `accounts` field of the `instruction`. When the instruction has duplicated
+/// accounts, it is necessary to pass a duplicated reference to the same account
+/// to maintain the 1:1 relationship between `account_infos` and `accounts`.
+#[inline(always)]
+pub fn invoke_signed_with_bounds<const MAX_ACCOUNTS: usize>(
     instruction: &Instruction,
     account_infos: &[&AccountInfo],
     signers_seeds: &[Signer],
 ) -> ProgramResult {
-    if instruction.accounts.len() < account_infos.len() {
-        return Err(ProgramError::NotEnoughAccountKeys);
+    // Check that the number of `MAX_ACCOUNTS` provided is not greater than
+    // the maximum number of accounts allowed.
+    const {
+        assert!(
+            MAX_ACCOUNTS <= MAX_CPI_ACCOUNTS,
+            "MAX_ACCOUNTS is greater than allowed MAX_CPI_ACCOUNTS"
+        );
     }
 
-    if account_infos.len() > MAX_CPI_ACCOUNTS {
-        return Err(ProgramError::InvalidArgument);
+    // Check that the number of accounts provided is not less than
+    // the number of accounts expected by the instruction.
+    if account_infos.len() < instruction.accounts.len() {
+        return Err(ProgramError::NotEnoughAccountKeys);
     }
 
     const UNINIT: MaybeUninit<Account> = MaybeUninit::<Account>::uninit();
-    let mut accounts = [UNINIT; MAX_CPI_ACCOUNTS];
-    let mut len = 0;
-
-    for (account_info, account_meta) in account_infos.iter().zip(instruction.accounts.iter()) {
-        if account_info.key() != account_meta.pubkey {
-            return Err(ProgramError::InvalidArgument);
-        }
-
-        let state = if account_meta.is_writable {
-            BorrowState::Borrowed
-        } else {
-            BorrowState::MutablyBorrowed
-        };
-
-        if account_info.is_borrowed(state) {
-            return Err(ProgramError::AccountBorrowFailed);
-        }
-
-        // SAFETY: The number of accounts has been validated to be less than
-        // `MAX_CPI_ACCOUNTS`.
-        unsafe {
-            accounts
-                .get_unchecked_mut(len)
-                .write(Account::from(*account_info));
-        }
-
-        len += 1;
-    }
-    // SAFETY: The accounts have been validated.
+    let mut accounts = [UNINIT; MAX_ACCOUNTS];
```

Comment:
If we know that the `Account` structures are so big, is there a good reason we do not allocate them on the heap?

```suggestion
    let mut accounts = Box::new([UNINIT; MAX_ACCOUNTS]);
```

CUs for the heap allocation are significant?